Feature/custom layouts

micHar opened the pull request:
The existing decorator pattern for composing custom day view wasn't enough for my needs. I wanted a custom layout where I could add e.g. an ImageView. I implemented it changing the CalendarCellView from TextView to FrameLayout and making a DayViewAdapter which can be used (via setCustomDayView) to inflate custom layout in lieu of the TextView. There is also a default implementation of this adapter, so that the basic experience doesn't change. I also added new sample.

I would like to go even further. E.g. make the default square cells optional so I can make layouts like this: https://s3-eu-west-1.amazonaws.com/uploads-eu.hipchat.com/83322/1764739/LfK3hqFNB2x1yQG/3.0%20Booking%402x.png. But I want to hear your thoughts first.

edenman commented:
@micHar thanks for the PR.  High-level things:

- I'm worried about the potential performance impact of the extra level of view hierarchy.  Can you please run some before/after timings for `MonthView`.`onMeasure`/`onLayout`/`onDraw`?
- I think if you need to get to something as custom as that design comp, it might be better to just fork.  Making this library generic enough to support your needs may affect usability/performance too much to be worth the effort.

Low-level things: 

The build is red, lots of checkstyle issues.  lmk when you've addressed those and I'll take a pass through the updated code.  Assuming the performance isn't an issue and you're still pushing ahead.  Further, if we do end up accepting this, you'll need to sign our (CLA)[https://docs.google.com/a/squareup.com/forms/d/13WR8m5uZ2nAkJH41k7GdVBXAAbzDk00vxtEYjd6Imzg/viewform?formkey=dDViT2xzUHAwRkI3X3k5Z0lQM091OGc6MQ] and squash all your commits.

micHar commented:
Thanks @edenman. I'll do these as soon as I finish my own work. Can you tell me how do I run the checkstyle without making the PR? I don't want to spam your PRs section

micHar commented:
I don't know if that's what you had in mind, but I have done something like this:

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    long pre = System.currentTimeMillis();
    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
    long post = System.currentTimeMillis();
    Log.d("MonthView", "onMeasure time: "+(post-pre)+" ms");
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
      long pre = System.currentTimeMillis();
      super.onLayout(changed, l, t, r, b);
      long post = System.currentTimeMillis();
      Log.d("MonthView", "onLayout time: "+(post-pre)+" ms");
    }

    @Override
    protected void onDraw(Canvas canvas) {
      long pre = System.currentTimeMillis();
      super.onDraw(canvas);
      long post = System.currentTimeMillis();
      Log.d("MonthView", "onDraw time: "+(post-pre)+" ms");
    }

onDraw doesn't get called in ViewGroups (as I have just learnt).
The rest of the test goes as follows:
Before:
- onMeasure (first time): 14ms.
- onLayout (first time): 2ms
- onMeasure (later, e.g. when scrolling the listview): 0-1ms
- onLayout (later): 0-1ms

After:
- onMeasure (first time): 20ms.
- onLayout (first time): 2ms
- onMeasure (later, e.g. when scrolling the listview): 0-1ms
- onLayout (later): 0-1ms

I also cleaned the code style and fixed one unit test (which assumed that CalendarCellView was a TextView). Now it passes the test.

Apart from that I also added optional square cells (via xml attribute). It doesn't add complexity to the API as it's optional and the default square cell shape doesn't change. If you set it to false it just becomes wrap_content. But this feature is on another branch and I can make a separate PR for this https://github.com/micHar/android-times-square/tree/feature/non-square-cells.

edenman commented:
Cool, those perf numbers seem fine to me.  Was mostly just concerned about scrolling performance.

Definitely keep the non-square-cells change separate.

edenman commented on library/src/main/java/com/squareup/timessquare/CalendarCellView.java:
```diff
@@ -36,6 +37,8 @@
   private boolean isHighlighted = false;
   private RangeState rangeState = RangeState.NONE;
 
```
remove this blank line

edenman commented on library/src/main/java/com/squareup/timessquare/CalendarCellView.java:
```diff
@@ -107,4 +110,18 @@ public boolean isSelectable() {
+
+  public TextView getDayOfMonthTextView() {
+    if (dayOfMonthTextView == null) {
+      throw new IllegalStateException(
```
nit:
```
      throw new IllegalStateException(
          "You have to setDayMonthOfView in your custom DayViewAdapter.");
```

edenman commented on library/src/main/java/com/squareup/timessquare/CalendarCellView.java:
```diff
@@ -107,4 +110,18 @@ public boolean isSelectable() {
+    }
+    return dayOfMonthTextView;
+  }
+
```
remove this blank line

edenman commented on library/src/main/java/com/squareup/timessquare/CalendarRowView.java:
```diff
@@ -79,6 +79,15 @@ public void setListener(MonthView.Listener listener) {
 
+  public void setDayViewAdapter(DayViewAdapter adapter) {
+    for (int i = 0; i < getChildCount(); i++) {
+      if (getChildAt(i) instanceof CalendarCellView) {
```
```
if (getChildAt(i) instanceof CalendarCellView) {
  CalendarCellView cell = ((CalendarCellView) getChildAt(i));
  cell.removeAllViews();
  adapter.makeCellView(cell);
}
```

edenman commented on library/src/main/java/com/squareup/timessquare/DayViewAdapter.java:
```diff
@@ -0,0 +1,10 @@
+ */
+public interface DayViewAdapter {
+
+    void makeCellView(CalendarCellView parent);
```
two-character indent.  remove the two blank lines.  remove the current class comment and add one explaining what this is used for.

edenman commented on library/src/main/java/com/squareup/timessquare/DefaultDayViewAdapter.java:
```diff
@@ -0,0 +1,20 @@
+ * Created by Michal on 30/11/15.
+ */
+public class DefaultDayViewAdapter implements DayViewAdapter {
+
```
again, two-char indent, no blank lines, either add a useful class comment or just remove it entirely.

edenman commented on library/src/main/res/values/strings.xml:
```diff
@@ -5,4 +5,5 @@
   <string name="day_name_format">EEE</string>
   <string name="invalid_date">Date must be between %1$s and %2$s.</string>
   <string name="month_name_format">MMMM yyyy</string>
+  <item type="id" name="day_view_adapter_class"/>
```
This should go in a new `ids.xml` file instead of in `strings.xml`

edenman commented on sample/src/main/java/com/squareup/timessquare/sample/SampleDayViewAdapter.java:
```diff
@@ -0,0 +1,24 @@
+ * Created by Michal on 01/12/15.
+ */
+public class SampleDayViewAdapter implements DayViewAdapter {
+
```
2-char indent, no blank lines, remove class comment

edenman commented on sample/src/main/res/layout/day_view_custom.xml:
```diff
@@ -0,0 +1,20 @@
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical"
+    android:padding="4dp">
```
`>` on the next line

edenman commented on sample/src/main/res/layout/day_view_custom.xml:
```diff
@@ -0,0 +1,20 @@
+        android:id="@+id/day_view"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        style="@style/CalendarCell.CalendarDate"/>
```
`/>` on the next line here and on line 18

edenman replied:
nm, looks like this is correct style for these files, ignore this one.

edenman commented on sample/src/main/res/layout/day_view_custom.xml:
```diff
@@ -0,0 +1,20 @@
+        android:layout_height="20dp"
+        android:src="@drawable/icon"
+        android:scaleType="centerInside"/>
+
```
remove blank line

edenman commented:
@micHar just a bunch of nits, LGTM.  Clean up the nits and squash your commits and then I think we'll be good to go!

micHar commented:
@edenman I'm really glad, thanks! I don't have much time to do it now, though. I will probably find some time to fix all these in the beginning of the next week.

micHar commented:
Hey @edenman, I finally found a moment to fix this. I hope everything is ready to merge now. Let me know if there's any problem.

edenmanjune commented on .travis.yml:
```diff
@@ -14,8 +14,8 @@ after_success:
   global:
-    - secure: "jpcnT7u80SX1SngObG41hZ0OarA3bje69YAV9t8daYCtNpu/RHoNjYhBmpCBZoEuL0zst7r5CmTkFSAf/V7CXnuUaYfeKCPo1LLZDBYIGgUYutGSJ1JkEby3eytek4HevG8hsjMOrhdQT27qyZ3ESNvW8JgpLd/R06/64SLbhWM="
-    - secure: "paXM9hwEIfdDo25fxLlVj9UbKPBx5HvOz/ehRr150SBxY9XhwXV3PkcII483vlmYiSE0yKKanSf5CDpteV1wRnQPUxF4QzyXmcwiDektV6ld7/qHRXWeZ89Bt3S8GZ5MEZSDhF6cmq5Ivi6v/5gvlobnTw1Vp55idbJZ94EdE7g="
+    - secure: "mokP2uluF4Yeg7ZvhjVOoo6E5lXWWyhhqGMZ4EbZ3xbl8W/a6oObwoLkVhxaMpdFz597VfBdiC0wX+y9ri2HBY2g2CCf2B4FFNsPFzA9MAKchFwfeqd/Z5GxfZVvEdYdSSueIRpeNJnjVVxt3WApCtNn7wsu9j3P4AtD76KxGL0="
```
what's going on here?

edenman replied:
what's going on here?

micHar replied:
No idea. I have whole git history on a separate branch, will check this out tommorow, but it must be some automatic ide change, cause I never touched Travis files

micHar replied:
I have no idea what it is. But the "mokP2ulu..." part is also on your master branch, so I thinks it's safe to go.

edenman replied:
Weird that it showed up in the diff.  Whevs!

edenman commented:
Thanks @micHar!
